File: go/pkg/minimax/http.go
package minimax

import (
	"bufio"
	"bytes"
	"context"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"strings"
	"time"
)

// httpClient handles HTTP communication with the MiniMax API.
type httpClient struct {
	client     *http.Client
	baseURL    string
	apiKey     string
	maxRetries int
}

// newHTTPClient creates a new HTTP client.
func newHTTPClient(cfg *clientConfig) *httpClient {
	return &httpClient{
		client:     cfg.httpClient,
		baseURL:    cfg.baseURL,
		apiKey:     cfg.apiKey,
		maxRetries: cfg.maxRetries,
	}
}

// apiResponse is the common response wrapper from MiniMax API.
type apiResponse struct {
	BaseResp *baseResp       `json:"base_resp,omitempty"`
	Data     json.RawMessage `json:"-"` // Will be unmarshaled separately
}

type baseResp struct {
	StatusCode int    `json:"status_code"`
	StatusMsg  string `json:"status_msg"`
}

// request makes an HTTP request to the API with retry support.
func (h *httpClient) request(ctx context.Context, method, path string, body any, result any) error {
	var bodyData []byte
	if body != nil {
		var err error
		bodyData, err = json.Marshal(body)
		if err != nil {
			return fmt.Errorf("marshal request body: %w", err)
		}
	}

	var lastErr error
	for attempt := 0; attempt <= h.maxRetries; attempt++ {
		if attempt > 0 {
			// Exponential backoff: 1s, 2s, 4s, ...
			backoff := time.Duration(1<<uint(attempt-1)) * time.Second
			select {
			case <-ctx.Done():
				return ctx.Err()
			case <-time.After(backoff):
			}
		}

		err := h.doRequest(ctx, method, path, bodyData, result)
		if err == nil {
			return nil
		}

		lastErr = err

		// Check if error is retryable
		if apiErr, ok := AsError(err); ok {
			if !apiErr.Retryable() {
				return err
			}
		} else {
			// Non-API errors (network errors) are retryable
			continue
		}
	}

	return lastErr
}

// doRequest performs a single HTTP request.
func (h *httpClient) doRequest(ctx context.Context, method, path string, bodyData []byte, result any) error {
	url := h.baseURL + path

	var bodyReader io.Reader
	if bodyData != nil {
		bodyReader = bytes.NewReader(bodyData)
	}

	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
	if err != nil {
		return fmt.Errorf("create request: %w", err)
	}

	h.setHeaders(req)
	if bodyData != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := h.client.Do(req)
	if err != nil {
		return fmt.Errorf("do request: %w", err)
	}
	defer resp.Body.Close()

	return h.handleResponse(resp, result)
}

// requestStream makes a streaming HTTP request to the API.
func (h *httpClient) requestStream(ctx context.Context, method, path string, body any) (*http.Response, error) {
	url := h.baseURL + path

	var bodyReader io.Reader
	if body != nil {
		data, err := json.Marshal(body)
		if err != nil {
			return nil, fmt.Errorf("marshal request body: %w", err)
		}
		bodyReader = bytes.NewReader(data)
	}

	req, err := http.NewRequestWithContext(ctx, method, url, bodyReader)
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}

	h.setHeaders(req)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	req.Header.Set("Accept", "text/event-stream")

	resp, err := h.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("do request: %w", err)
	}

	// Check for error response (non-streaming)
	if resp.StatusCode != http.StatusOK {
		defer resp.Body.Close()
		return nil, h.handleErrorResponse(resp)
	}

	return resp, nil
}

// uploadFile uploads a file using multipart form data with streaming.
// This avoids loading the entire file into memory.
func (h *httpClient) uploadFile(ctx context.Context, path string, file io.Reader, filename string, fields map[string]string, result any) error {
	url := h.baseURL + path

	// Use io.Pipe for streaming upload to avoid loading entire file into memory
	pr, pw := io.Pipe()
	writer := multipart.NewWriter(pw)

	// Write multipart data in a goroutine
	errCh := make(chan error, 1)
	go func() {
		defer pw.Close()

		// Add file field
		part, err := writer.CreateFormFile("file", filename)
		if err != nil {
			errCh <- fmt.Errorf("create form file: %w", err)
			return
		}
		if _, err := io.Copy(part, file); err != nil {
			errCh <- fmt.Errorf("copy file: %w", err)
			return
		}

		// Add other fields
		for key, value := range fields {
			if err := writer.WriteField(key, value); err != nil {
				errCh <- fmt.Errorf("write field %s: %w", key, err)
				return
			}
		}

		if err := writer.Close(); err != nil {
			errCh <- fmt.Errorf("close writer: %w", err)
			return
		}

		errCh <- nil
	}()

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, pr)
	if err != nil {
		pr.Close()
		return fmt.Errorf("create request: %w", err)
	}

	h.setHeaders(req)
	req.Header.Set("Content-Type", writer.FormDataContentType())

	resp, err := h.client.Do(req)
	if err != nil {
		return fmt.Errorf("do request: %w", err)
	}
	defer resp.Body.Close()

	// Check for errors from the goroutine
	if writeErr := <-errCh; writeErr != nil {
		return writeErr
	}

	return h.handleResponse(resp, result)
}

// setHeaders sets common headers for API requests.
func (h *httpClient) setHeaders(req *http.Request) {
	req.Header.Set("Authorization", "Bearer "+h.apiKey)
	req.Header.Set("User-Agent", "giztoy-minimax-go/1.0")
}

// handleResponse handles the API response.
func (h *httpClient) handleResponse(resp *http.Response, result any) error {
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("read response body: %w", err)
	}

	// Check HTTP status
	if resp.StatusCode != http.StatusOK {
		return h.parseError(body, resp.StatusCode)
	}

	// First check for API-level error in base_resp
	var apiResp apiResponse
	if err := json.Unmarshal(body, &apiResp); err == nil {
		if apiResp.BaseResp != nil && apiResp.BaseResp.StatusCode != 0 {
			return &Error{
				StatusCode: apiResp.BaseResp.StatusCode,
				StatusMsg:  apiResp.BaseResp.StatusMsg,
				HTTPStatus: resp.StatusCode,
			}
		}
	}

	// Parse response into result if provided
	if result != nil {
		if err := json.Unmarshal(body, result); err != nil {
			return fmt.Errorf("unmarshal response: %w", err)
		}
	}

	return nil
}

// handleErrorResponse handles an error response.
func (h *httpClient) handleErrorResponse(resp *http.Response) error {
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return fmt.Errorf("read error response: %w", err)
	}
	return h.parseError(body, resp.StatusCode)
}

// parseError parses an error response body.
func (h *httpClient) parseError(body []byte, httpStatus int) error {
	var apiResp struct {
		BaseResp *baseResp `json:"base_resp"`
	}
	if err := json.Unmarshal(body, &apiResp); err == nil && apiResp.BaseResp != nil {
		return &Error{
			StatusCode: apiResp.BaseResp.StatusCode,
			StatusMsg:  apiResp.BaseResp.StatusMsg,
			HTTPStatus: httpStatus,
		}
	}

	return &Error{
		StatusCode: httpStatus,
		StatusMsg:  string(body),
		HTTPStatus: httpStatus,
	}
}

// sseReader helps read Server-Sent Events from a response.
type sseReader struct {
	reader  *bufio.Reader
	resp    *http.Response
	onClose func()
}

// newSSEReader creates a new SSE reader.
func newSSEReader(resp *http.Response) *sseReader {
	return &sseReader{
		reader: bufio.NewReader(resp.Body),
		resp:   resp,
	}
}

// readEvent reads the next SSE event.
// Returns (data, isDone, error).
func (r *sseReader) readEvent() ([]byte, bool, error) {
	var data []byte

	for {
		line, err := r.reader.ReadBytes('\n')
		if err != nil {
			if err == io.EOF {
				return nil, true, nil
			}
			return nil, false, err
		}

		line = bytes.TrimSpace(line)
		if len(line) == 0 {
			// Empty line marks end of event
			if len(data) > 0 {
				return data, false, nil
			}
			continue
		}

		if bytes.HasPrefix(line, []byte("data:")) {
			eventData := bytes.TrimPrefix(line, []byte("data:"))
			eventData = bytes.TrimSpace(eventData)

			// Check for stream end marker
			if bytes.Equal(eventData, []byte("[DONE]")) {
				return nil, true, nil
			}

			data = eventData
		}
	}
}

// close closes the SSE reader.
func (r *sseReader) close() {
	r.resp.Body.Close()
	if r.onClose != nil {
		r.onClose()
	}
}

// decodeHexAudio decodes hex-encoded audio data.
func decodeHexAudio(hexData string) ([]byte, error) {
	// Remove any whitespace
	hexData = strings.ReplaceAll(hexData, " ", "")
	hexData = strings.ReplaceAll(hexData, "\n", "")

	return hex.DecodeString(hexData)
}

File: rust/minimax/src/http.rs
//! HTTP client implementation for MiniMax API.

use std::time::Duration;

use bytes::Bytes;
use futures::{Stream, StreamExt};
use reqwest::{
    header::{HeaderMap, HeaderValue, AUTHORIZATION, CONTENT_TYPE, USER_AGENT},
    multipart, Client as ReqwestClient, Response, StatusCode,
};
use serde::{de::DeserializeOwned, Serialize};

use super::{
    error::{Error, Result},
    types::BaseResp,
};

/// HTTP client for MiniMax API.
pub struct HttpClient {
    client: ReqwestClient,
    base_url: String,
    api_key: String,
    max_retries: u32,
}

impl HttpClient {
    /// Creates a new HTTP client.
    pub fn new(base_url: String, api_key: String, max_retries: u32) -> Result<Self> {
        let client = ReqwestClient::builder()
            .timeout(Duration::from_secs(300))
            .build()?;

        Ok(Self {
            client,
            base_url,
            api_key,
            max_retries,
        })
    }

    /// Makes an HTTP request to the API with retry support.
    pub async fn request<T, R>(&self, method: &str, path: &str, body: Option<&T>) -> Result<R>
    where
        T: Serialize + ?Sized,
        R: DeserializeOwned,
    {
        let mut last_err = None;

        for attempt in 0..=self.max_retries {
            if attempt > 0 {
                // Exponential backoff: 1s, 2s, 4s, ...
                let backoff = Duration::from_secs(1 << (attempt - 1));
                tokio::time::sleep(backoff).await;
            }

            match self.do_request(method, path, body).await {
                Ok(result) => return Ok(result),
                Err(e) => {
                    if e.is_retryable() {
                        last_err = Some(e);
                        continue;
                    }
                    return Err(e);
                }
            }
        }

        Err(last_err.unwrap_or_else(|| Error::Other("max retries exceeded".to_string())))
    }

    /// Performs a single HTTP request.
    async fn do_request<T, R>(&self, method: &str, path: &str, body: Option<&T>) -> Result<R>
    where
        T: Serialize + ?Sized,
        R: DeserializeOwned,
    {
        let url = format!("{}{}", self.base_url, path);

        let mut request = match method {
            "GET" => self.client.get(&url),
            "POST" => self.client.post(&url),
            "PUT" => self.client.put(&url),
            "DELETE" => self.client.delete(&url),
            _ => return Err(Error::Other(format!("unsupported method: {}", method))),
        };

        request = request.headers(self.default_headers());

        if let Some(body) = body {
            request = request.json(body);
        }

        let response = request.send().await?;
        self.handle_response(response).await
    }

    /// Makes a streaming HTTP request.
    pub async fn request_stream<T>(
        &self,
        method: &str,
        path: &str,
        body: Option<T>,
    ) -> Result<impl Stream<Item = Result<Bytes>> + use<T>>
    where
        T: Serialize,
    {
        let url = format!("{}{}", self.base_url, path);

        let mut request = match method {
            "GET" => self.client.get(&url),
            "POST" => self.client.post(&url),
            _ => return Err(Error::Other(format!("unsupported method: {}", method))),
        };

        let mut headers = self.default_headers();
        headers.insert("Accept", HeaderValue::from_static("text/event-stream"));
        request = request.headers(headers);

        if let Some(ref body) = body {
            request = request.json(body);
        }

        let response = request.send().await?;

        if response.status() != StatusCode::OK {
            return Err(self.handle_error_response(response).await);
        }

        Ok(response.bytes_stream().map(|r| r.map_err(Error::from)))
    }

    /// Uploads a file using multipart form data.
    pub async fn upload_file<R>(
        &self,
        path: &str,
        file_bytes: Vec<u8>,
        filename: &str,
        fields: Vec<(&str, String)>,
    ) -> Result<R>
    where
        R: DeserializeOwned,
    {
        let url = format!("{}{}", self.base_url, path);

        let mut form = multipart::Form::new().part(
            "file",
            multipart::Part::bytes(file_bytes).file_name(filename.to_string()),
        );

        for (key, value) in fields {
            form = form.text(key.to_string(), value);
        }

        let mut headers = HeaderMap::new();
        headers.insert(
            AUTHORIZATION,
            HeaderValue::from_str(&format!("Bearer {}", self.api_key))
                .map_err(|e| Error::Other(e.to_string()))?,
        );
        headers.insert(
            USER_AGENT,
            HeaderValue::from_static("giztoy-minimax-rust/1.0"),
        );

        let response = self
            .client
            .post(&url)
            .headers(headers)
            .multipart(form)
            .send()
            .await?;

        self.handle_response(response).await
    }

    /// Returns default headers for API requests.
    fn default_headers(&self) -> HeaderMap {
        let mut headers = HeaderMap::new();
        headers.insert(
            AUTHORIZATION,
            HeaderValue::from_str(&format!("Bearer {}", self.api_key)).unwrap(),
        );
        headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
        headers.insert(
            USER_AGENT,
            HeaderValue::from_static("giztoy-minimax-rust/1.0"),
        );
        headers
    }

    /// Handles the API response.
    async fn handle_response<R>(&self, response: Response) -> Result<R>
    where
        R: DeserializeOwned,
    {
        let status = response.status();
        let body = response.bytes().await?;

        if !status.is_success() {
            return Err(self.parse_error(&body, status.as_u16()));
        }

        // Check for API-level error in base_resp
        if let Ok(api_resp) = serde_json::from_slice::<ApiResponse>(&body) {
            if let Some(base_resp) = api_resp.base_resp {
                if base_resp.is_error() {
                    return Err(Error::api(
                        base_resp.status_code,
                        base_resp.status_msg,
                        status.as_u16(),
                    ));
                }
            }
        }

        serde_json::from_slice(&body).map_err(Error::from)
    }

    /// Handles an error response.
    async fn handle_error_response(&self, response: Response) -> Error {
        let status = response.status().as_u16();
        match response.bytes().await {
            Ok(body) => self.parse_error(&body, status),
            Err(e) => Error::Http(e),
        }
    }

    /// Parses an error response body.
    fn parse_error(&self, body: &[u8], http_status: u16) -> Error {
        if let Ok(api_resp) = serde_json::from_slice::<ApiResponse>(body) {
            if let Some(base_resp) = api_resp.base_resp {
                return Error::api(base_resp.status_code, base_resp.status_msg, http_status);
            }
        }

        Error::api(
            http_status as i32,
            String::from_utf8_lossy(body).to_string(),
            http_status,
        )
    }
}

/// API response wrapper.
#[derive(Debug, serde::Deserialize)]
struct ApiResponse {
    base_resp: Option<BaseResp>,
}

/// SSE (Server-Sent Events) reader.
pub(crate) struct SseReader<S> {
    stream: S,
    buffer: String,
}

impl<S> SseReader<S>
where
    S: Stream<Item = Result<Bytes>> + Unpin,
{
    pub fn new(stream: S) -> Self {
        Self {
            stream,
            buffer: String::new(),
        }
    }

    /// Reads the next SSE event.
    /// Returns (data, is_done).
    pub async fn read_event(&mut self) -> Result<Option<Vec<u8>>> {
        loop {
            // Check if we have a complete event in the buffer
            if let Some(event) = self.extract_event() {
                if event == "[DONE]" {
                    return Ok(None);
                }
                return Ok(Some(event.into_bytes()));
            }

            // Read more data from the stream
            match self.stream.next().await {
                Some(Ok(bytes)) => {
                    self.buffer.push_str(&String::from_utf8_lossy(&bytes));
                }
                Some(Err(e)) => return Err(e),
                None => return Ok(None),
            }
        }
    }

    /// Extracts a complete event from the buffer.
    fn extract_event(&mut self) -> Option<String> {
        // Find "data:" line followed by a double newline
        let mut search_pos = 0;

        while let Some(pos) = self.buffer[search_pos..].find("data:") {
            let abs_pos = search_pos + pos;
            let after_data = abs_pos + 5; // "data:" length

            // Skip whitespace after "data:"
            let content_start = self.buffer[after_data..]
                .chars()
                .take_while(|c| *c == ' ')
                .count()
                + after_data;

            // Find the end of this data line
            if let Some(newline_pos) = self.buffer[content_start..].find('\n') {
                let content_end = content_start + newline_pos;
                let content = self.buffer[content_start..content_end].trim();

                // Check if this is a complete event (followed by empty line or another event)
                let after_newline = content_end + 1;
                if after_newline >= self.buffer.len()
                    || self.buffer[after_newline..].starts_with('\n')
                    || self.buffer[after_newline..].starts_with("data:")
                {
                    let result = content.to_string();
                    self.buffer = self.buffer[after_newline..].trim_start_matches('\n').to_string();
                    return Some(result);
                }
            }

            search_pos = abs_pos + 1;
        }

        None
    }
}

/// Decodes hex-encoded audio data.
pub(crate) fn decode_hex_audio(hex_data: &str) -> Result<Vec<u8>> {
    let cleaned: String = hex_data.chars().filter(|c| !c.is_whitespace()).collect();
    hex::decode(&cleaned).map_err(Error::from)
}
